File: akf_hrms/akf_hrms/report/mobile_allowance_details_report/mobile_allowance_details_report.js
// Developer Mubashir Bashir

frappe.query_reports["Mobile Allowance Details Report"] = {
	"filters": [
		{
			"fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname":"employee",
			"label": __("Employee"),
			"fieldtype": "Link",
			"options": "Employee",
			// "reqd": 1
		},
		{
			"fieldname":"designation",
			"label": __("Designation"),
			"fieldtype": "Link",
			"options": "Designation"
		},
		{
			"fieldname":"department",
			"label": __("Department"),
			"fieldtype": "Link",
			"options": "Department"
		},
		{
			"fieldname":"employment_type",
			"label": __("Employment Type"),
			"fieldtype": "Link",
			"options": "Employment Type"
		},
		{
			"fieldname":"grade",
			"label": __("Employee Grade"),
			"fieldtype": "Link",
			"options": "Employee Grade"
		},
		// {
		// 	"label": "Date",
		// 	"fieldtype": "Date",
		// 	"fieldname": "date",
		// 	"default": frappe.datetime.nowdate("dd-mm-yyyy"),
		// 	"reqd": 0,
		// },
	],	
	
	onload: function(report) {
		if (frappe.user.has_role("Employee")) {
			frappe.call({
				method: "frappe.client.get_value",
				args: {
					doctype: "Employee",
					fieldname: ["name", "designation", "department", "employment_type", "grade"],
					filters: { user_id: frappe.session.user }
				},
				callback: function(response) {
					if (response && response.message) {
						const employee_id = response.message.name;
						const designation = response.message.designation;
						const department = response.message.department;
						const grade = response.message.grade;
						const employment_type = response.message.employment_type;

						report.set_filter_value("employee", employee_id);
						if (designation) {
							report.set_filter_value("designation", designation);
						}
						if (department) {
							report.set_filter_value("department", department);
						}
						if (grade) {
							report.set_filter_value("grade", grade);
						}
						if (employment_type) {
							report.set_filter_value("employment_type", employment_type);
						}
					}
				}
			});
		}
	}
};

File: akf_hrms/akf_hrms/report/mobile_allowance_details_report/mobile_allowance_details_report.py
# Developer Mubashir Bashir

from __future__ import unicode_literals
from frappe import _
import frappe
import ast

def execute(filters=None):
	columns, data = [], []
	columns = get_columns()
	data = get_data(filters)	
	return columns, data

def get_columns():
	return [
		_("Employee ID") + ":Link/Employee:120",
		_("Employee name") + ":Data:180",
		_("Branch") + ":Data:180",
		_("Department") + ":Data:180",
		_("Designation") + ":Data:180",
		_("Type") + ":Data:110",
		_("Grade") + ":Data:90",
		_("Mobile Allowance") + ":Data:180",
		_("CNIC") + ":Data:120",
		_("Contact") + ":Data:120",
		_("Issue Date") + ":Date:120",
		_("Balance on Issue") + ":Data:120",
		_("Return Date") + ":Date:120",
		_("Balance on Return") + ":Data:120",
	]


def get_data(filters):
    conditions = get_condition(filters)    
    query = """ 
        SELECT
            e.name, e.employee_name, e.branch, e.department, e.designation, e.employment_type, e.grade,
            COALESCE(
                (SELECT custom_mobile_allowance 
                FROM `tabSalary Structure Assignment` ssa
                WHERE ssa.employee = e.name 
                AND ssa.docstatus = 1
                AND ssa.from_date <= CURDATE()
                ORDER BY ssa.from_date DESC
                LIMIT 1), 
                0
            ) as custom_mobile_allowence,
            e.custom_cnic, ah.sim_number, ah.issue_date, ah.balance_on_issue, ah.return_date, ah.balance_on_return
        FROM `tabEmployee` e
        LEFT JOIN `tabAsset History` ah ON ah.parent = e.name
        WHERE e.status = 'Active' {condition}
        ORDER BY e.name, ah.issue_date DESC
    """.format(condition=conditions)

    results = frappe.db.sql(query, filters, as_dict=1)
    
    def format_currency(value):
        if not value:
            return "Rs 0"
        return frappe.utils.fmt_money(value, currency='Rs').replace('.00', '')

    formatted_result = []
    previous_employee = None

    for row in results:
        if row.get("name") == previous_employee:
            formatted_row = ['-'] * 9
        else:
            formatted_row = [
                row.get("name"),
                row.get("employee_name"),
                row.get("branch"),
                row.get("department"),
                row.get("designation"),
                row.get("employment_type"),
                row.get("grade"),
                format_currency(row.get("custom_mobile_allowence")),
                row.get("custom_cnic"),                 
            ]
            previous_employee = row.get("name")
        
        formatted_row += [            
            row.get("sim_number"),
            row.get("issue_date"),
            format_currency(row.get("balance_on_issue")),
            row.get("return_date"),
            format_currency(row.get("balance_on_return"))
        ]

        formatted_result.append(formatted_row)
    return formatted_result

def get_condition(filters):
    conditions = ""
    if filters.get("company"):
        conditions += " and e.company = %(company)s"
    if filters.get("employee"):
        conditions += " AND e.employee = %(employee)s"
    if filters.get("designation"):
        conditions += " AND e.designation = %(designation)s"
    if filters.get("department"):
        conditions += " AND e.department = %(department)s"
    if filters.get("employment_type"):
        conditions += " AND e.employment_type = %(employment_type)s"
    if filters.get("grade"):
        conditions += " AND e.grade = %(grade)s"
    if filters.get("branch"):
        conditions += " AND e.branch = %(branch)s"
    
    return conditions  
